File: python_legacy/lib/epub_builder.py
import html
import io
import json
import logging
import os
import re
from os.path import exists, join, splitext
from typing import List, Optional, Tuple

from bs4 import BeautifulSoup
from ebooklib import epub
from PIL import Image, ImageFile

from .constants import RESULT_DIR
from .utils import TextUtils

# --- FIX FOR TRUNCATED IMAGES ---
ImageFile.LOAD_TRUNCATED_IMAGES = True
# -----

logger = logging.getLogger(__name__)


class EpubBuilder:
    def __init__(self, base_folder: str, compress_images: bool = True):
        self.base_folder = base_folder
        self.compress_images = compress_images
        self.image_map = {}
        self.result_root = RESULT_DIR

        self.meta = {}
        meta_path = join(base_folder, "metadata.json")
        if exists(meta_path):
            with open(meta_path, "r", encoding="utf-8") as f:
                self.meta = json.load(f)
        else:
            self.meta = {
                "novel_name": "Unknown",
                "author": "Unknown",
                "summary": "",
                "cover_image_local": "",
                "tags": [],
            }

        self.css = """
            body { margin: 0; padding: 5px; text-align: justify; line-height: 1.4em; font-family: serif; }
            h1, h2, h3 { text-align: center; margin: 1em 0; font-weight: bold; }
            img { display: block; margin: 10px auto; max-width: 100%; height: auto; }
            p { margin-bottom: 1em; text-indent: 1em; }
            .center { text-align: center; }
            nav#toc ol { list-style-type: none; padding-left: 0; }
            nav#toc > ol > li { margin-top: 1em; font-weight: bold; }
            nav#toc > ol > li > ol { list-style-type: none; padding-left: 1.5em; font-weight: normal; }
            nav#toc > ol > li > ol > li { margin-top: 0.5em; }
            nav#toc a { text-decoration: none; color: inherit; }
            nav#toc a:hover { text-decoration: underline; color: blue; }
            a.footnote-link { vertical-align: super; font-size: 0.75em; text-decoration: none; color: #007bff; margin-left: 2px; }
            aside.footnote-content { margin-top: 1em; padding: 0.5em; border-top: 1px solid #ccc; font-size: 0.9em; color: #333; background-color: #f9f9f9; display: block; }
            aside.footnote-content p { margin: 0; text-indent: 0; }
            aside.footnote-content div.note-header { font-weight: bold; margin-bottom: 0.5em; color: #555; }
        """

    def _get_output_path(
        self, filename: str, is_merged: bool, create_dirs: bool = True
    ) -> str:
        """
        Determines the final output path based on user rules.
        1. Merged + Original -> result/<BookName - Full>.epub
        2. Compressed (Any) -> result/<BookName>/compressed/<filename>
        3. Separate + Original -> result/<BookName>/original/<filename>
        """
        book_name_slug = TextUtils.format_filename(self.meta["novel_name"])

        # Case 1: Merged & Original (Special Case)
        if is_merged and not self.compress_images:
            if create_dirs and not exists(self.result_root):
                os.makedirs(self.result_root)
            return join(self.result_root, filename)

        # Determine subfolder
        subfolder = "compressed" if self.compress_images else "original"

        # Path: result/<BookName>/<subfolder>/
        target_dir = join(self.result_root, book_name_slug, subfolder)

        if create_dirs and not exists(target_dir):
            os.makedirs(target_dir)

        return join(target_dir, filename)

    def sanitize_xhtml(self, html_content: str) -> str:
        if not html_content:
            return ""
        safe = html_content
        safe = safe.replace("&nbsp;", "&#160;")
        pattern_empty_p = re.compile(
            r"<p[^>]*>(\s|&nbsp;|&#160;|<br\s*\/?>)*<\/p>", re.IGNORECASE
        )
        safe = pattern_empty_p.sub("", safe)
        pattern_br = re.compile(r"(<br\s*\/?>\s*){3,}", re.IGNORECASE)
        safe = pattern_br.sub("<br/><br/>", safe)
        return safe.strip()

    def process_image(self, rel_path: str) -> Tuple[Optional[epub.EpubItem], str]:
        if not rel_path:
            return None, ""

        if rel_path in self.image_map:
            return None, self.image_map[rel_path]

        full_path = join(self.base_folder, rel_path)
        if not exists(full_path):
            return None, rel_path

        try:
            with Image.open(full_path) as valid_check:
                valid_check.load()

            if not self.compress_images:
                with open(full_path, "rb") as f:
                    content = f.read()

                ext = splitext(rel_path)[1].lower()
                media_type = "image/jpeg"
                if ext == ".png":
                    media_type = "image/png"
                elif ext == ".gif":
                    media_type = "image/gif"
                elif ext == ".webp":
                    media_type = "image/webp"

                item = epub.EpubItem(
                    uid=rel_path.replace("/", "_").replace(".", "_"),
                    file_name=rel_path,
                    media_type=media_type,
                    content=content,
                )
                self.image_map[rel_path] = rel_path
                return item, rel_path

            img = Image.open(full_path)
            if img.mode in ("RGBA", "P"):
                img = img.convert("RGB")

            output = io.BytesIO()
            img.save(output, format="JPEG", quality=75, optimize=True)

            base, _ = splitext(rel_path)
            new_rel_path = f"{base}.jpg"

            item = epub.EpubItem(
                uid=new_rel_path.replace("/", "_").replace(".", "_"),
                file_name=new_rel_path,
                media_type="image/jpeg",
                content=output.getvalue(),
            )

            self.image_map[rel_path] = new_rel_path
            return item, new_rel_path

        except Exception as e:
            error_msg = str(e).lower()
            if "truncated" in error_msg or "cannot identify" in error_msg:
                logger.error(f"CORRUPT FILE FOUND: {rel_path}")
                logger.warning(f"-> Deleting {rel_path}...")
                try:
                    os.remove(full_path)
                    logger.warning(
                        "-> File deleted. Please run 'Download' action again to re-fetch it."
                    )
                except OSError:
                    logger.error("-> Could not delete file (locked?). Delete manually.")
                return None, ""

            logger.error(f"Image process failed for {rel_path}: {e}")
            return None, rel_path

    def make_intro(self, vol_name: str = ""):
        summary_html = self.sanitize_xhtml(self.meta.get("summary", ""))
        title = html.escape(self.meta["novel_name"])
        author = html.escape(self.meta["author"])
        tags_str = ", ".join(self.meta.get("tags", []))
        tags_html = f"<p><b>Thể loại:</b> {tags_str}</p>" if tags_str else ""

        cover_html = "<hr/>"
        main_cover_path = self.meta.get("cover_image_local")

        cover_item = None
        if main_cover_path:
            item, new_path = self.process_image(main_cover_path)
            cover_item = item
            if new_path:
                cover_html = f'<div style="text-align:center; margin: 2em 0; page-break-after: always; break-after: page;"><img src="{new_path}" alt="Cover"/></div>'

        content = f"""
            <div style="text-align: center; margin-top: 5%;">
                <h1>{title}</h1>
                <h3 style="margin-bottom: 0.5em;">{vol_name}</h3>
                <p><b>Tác giả:</b> {author}</p>
                {tags_html}
                {cover_html}
                <div style="text-align: justify;">
                    {summary_html}
                </div>
            </div>
        """
        page = epub.EpubHtml(
            title="Giới thiệu", file_name="intro.xhtml", content=content
        )
        return page, cover_item

    def build_merged(self, json_files: List[str]):
        if "volumes" in self.meta:
            order_map = {v["filename"]: v["order"] for v in self.meta["volumes"]}
            json_files.sort(key=lambda x: order_map.get(x, 9999))

        book = epub.EpubBook()
        book.set_title(f"{self.meta['novel_name']}")
        book.set_language("vi")
        book.add_author(self.meta["author"])

        if self.meta.get("summary"):
            book.add_metadata("DC", "description", self.meta["summary"])
        for tag in self.meta.get("tags", []):
            book.add_metadata("DC", "subject", tag)

        css = epub.EpubItem(
            uid="style", file_name="style.css", media_type="text/css", content=self.css
        )
        book.add_item(css)

        intro_page, main_cover_item = self.make_intro("Toàn tập")
        intro_page.add_item(css)

        if main_cover_item:
            book.add_item(main_cover_item)
            book.set_cover("cover.jpg", main_cover_item.content, create_page=False)

        book.add_item(intro_page)
        spine = [intro_page]
        toc = [epub.Link("intro.xhtml", "Giới thiệu", "intro")]

        for i, jf in enumerate(json_files):
            print(f"Merging: {jf}")
            with open(join(self.base_folder, jf), "r", encoding="utf-8") as f:
                vol_data = json.load(f)

            vol_html_content = ""
            vol_cover = vol_data.get("cover_image_local")

            if vol_cover:
                item, new_src = self.process_image(vol_cover)
                if item:
                    book.add_item(item)
                vol_html_content += (
                    f'<img src="{new_src}" alt="Vol Cover" style="max-height: 50vh;"/>'
                )

            vol_html_content += f"<h1>{html.escape(vol_data['volume_name'])}</h1>"

            sep_html = f"""
                <div style="text-align: center; margin-top: 30vh;">
                    {vol_html_content}
                </div>
            """

            sep_page = epub.EpubHtml(
                title=vol_data["volume_name"],
                file_name=f"vol_{i}.xhtml",
                content=sep_html,
            )
            sep_page.add_item(css)
            book.add_item(sep_page)
            spine.append(sep_page)

            vol_chaps = []
            for chap in vol_data["chapters"]:
                soup = BeautifulSoup(chap["content"], "html.parser")
                for img in soup.find_all("img"):
                    src = img.get("src")
                    if src:
                        item, new_src = self.process_image(src)
                        if item:
                            book.add_item(item)
                        img["src"] = new_src

                clean_content = self.sanitize_xhtml(str(soup))
                fname = f"v{i}_c{chap['index']}.xhtml"
                c_page = epub.EpubHtml(
                    title=chap["title"],
                    file_name=fname,
                    content=f"<h2>{html.escape(chap['title'])}</h2>{clean_content}",
                )
                c_page.add_item(css)
                book.add_item(c_page)
                vol_chaps.append(c_page)

            spine.extend(vol_chaps)
            toc.append((sep_page, vol_chaps))

        nav = epub.EpubNav()
        book.add_item(nav)
        book.spine = ["nav"] + spine
        book.toc = toc
        book.add_item(epub.EpubNcx())

        base_filename = TextUtils.format_filename(f"{self.meta['novel_name']} Full")
        filename = f"{base_filename}.epub"
        out_path = self._get_output_path(filename, is_merged=True)

        epub.write_epub(out_path, book, {})
        print(f"Created Merged EPUB: {out_path}")

    def build_volume(self, json_file: str):
        with open(join(self.base_folder, json_file), "r", encoding="utf-8") as f:
            vol_data = json.load(f)

        book = epub.EpubBook()
        book.set_title(f"{vol_data['volume_name']} - {self.meta['novel_name']}")
        book.set_language("vi")
        book.add_author(self.meta["author"])
        for tag in self.meta.get("tags", []):
            book.add_metadata("DC", "subject", tag)

        css = epub.EpubItem(
            uid="style", file_name="style.css", media_type="text/css", content=self.css
        )
        book.add_item(css)

        intro_page, main_cover_item = self.make_intro(vol_data["volume_name"])
        intro_page.add_item(css)
        if main_cover_item:
            book.add_item(main_cover_item)
            book.set_cover("cover.jpg", main_cover_item.content, create_page=False)
        book.add_item(intro_page)

        spine = [intro_page]

        for chap in vol_data["chapters"]:
            soup = BeautifulSoup(chap["content"], "html.parser")
            for img in soup.find_all("img"):
                src = img.get("src")
                if src:
                    item, new_src = self.process_image(src)
                    if item:
                        book.add_item(item)
                    img["src"] = new_src

            clean = self.sanitize_xhtml(str(soup))
            c_page = epub.EpubHtml(
                title=chap["title"],
                file_name=f"c{chap['index']}.xhtml",
                content=f"<h2>{html.escape(chap['title'])}</h2>{clean}",
            )
            c_page.add_item(css)
            book.add_item(c_page)
            spine.append(c_page)

        nav = epub.EpubNav()
        book.add_item(nav)
        book.spine = ["nav"] + spine
        book.add_item(epub.EpubNcx())

        base_filename = TextUtils.format_filename(
            f"{vol_data['volume_name']} - {self.meta['novel_name']}"
        )
        filename = f"{base_filename}.epub"
        out_path = self._get_output_path(filename, is_merged=False)

        epub.write_epub(out_path, book, {})
        print(f"Created: {out_path}")

File: src/services/epub-builder.ts
/**
 * EPUB Builder Service
 * Requirements: 6.1, 6.2, 6.3, 6.4, 6.5, 6.6, 6.7, 6.8
 */

import { join, extname } from 'node:path';
import { readFile, writeFile } from 'node:fs/promises';
import epub, { type Chapter as EpubChapter, type Options as EpubGenOptions } from 'epub-gen-memory';
import * as cheerio from 'cheerio';
import sharp from 'sharp';

import type { NovelMetadata, VolumeData, EpubOptions as BuilderOptions } from '../types';
import { ensureDir, readJson, fileExistsWithContent } from '../utils/fs';
import { formatFilename } from '../utils/text';
import { PATHS } from '../config/constants';

/**
 * Processed image result
 */
interface ProcessedImage {
    data: Buffer;
    mimeType: string;
    newPath: string;
}

/**
 * CSS styling for EPUB content
 */
const EPUB_CSS = `
body { margin: 0; padding: 5px; text-align: justify; line-height: 1.4em; font-family: serif; }
h1, h2, h3 { text-align: center; margin: 1em 0; font-weight: bold; }
img { display: block; margin: 10px auto; max-width: 100%; height: auto; }
p { margin-bottom: 1em; text-indent: 1em; }
.center { text-align: center; }
nav#toc ol { list-style-type: none; padding-left: 0; }
nav#toc > ol > li { margin-top: 1em; font-weight: bold; }
nav#toc > ol > li > ol { list-style-type: none; padding-left: 1.5em; font-weight: normal; }
nav#toc > ol > li > ol > li { margin-top: 0.5em; }
nav#toc a { text-decoration: none; color: inherit; }
nav#toc a:hover { text-decoration: underline; color: blue; }
a.footnote-link { vertical-align: super; font-size: 0.75em; text-decoration: none; color: #007bff; margin-left: 2px; }
aside.footnote-content { margin-top: 1em; padding: 0.5em; border-top: 1px solid #ccc; font-size: 0.9em; color: #333; background-color: #f9f9f9; display: block; }
aside.footnote-content p { margin: 0; text-indent: 0; }
aside.footnote-content div.note-header { font-weight: bold; margin-bottom: 0.5em; color: #555; }
`.trim();

/**
 * EpubBuilder class for generating EPUB files from downloaded novel data
 */
export class EpubBuilder {
    private baseFolder: string;
    private compressImages: boolean;
    private outputDir: string;
    private metadata: NovelMetadata | null = null;
    private imageCache: Map<string, ProcessedImage> = new Map();

    constructor(baseFolder: string, options: BuilderOptions) {
        this.baseFolder = baseFolder;
        this.compressImages = options.compressImages;
        this.outputDir = options.outputDir || PATHS.RESULT_DIR;
    }

    /**
     * Load metadata from the base folder
     */
    private async loadMetadata(): Promise<NovelMetadata> {
        if (this.metadata) {
            return this.metadata;
        }

        const metaPath = join(this.baseFolder, 'metadata.json');
        try {
            this.metadata = await readJson<NovelMetadata>(metaPath);
            return this.metadata;
        } catch {
            // Return default metadata if file doesn't exist
            this.metadata = {
                novelName: 'Unknown',
                author: 'Unknown',
                summary: '',
                coverImageLocal: '',
                tags: [],
                url: '',
                volumes: [],
            };
            return this.metadata;
        }
    }


    /**
     * Load volume data from a JSON file
     */
    private async loadVolumeData(jsonFile: string): Promise<VolumeData> {
        const filePath = join(this.baseFolder, jsonFile);
        return readJson<VolumeData>(filePath);
    }

    /**
     * Sanitize HTML content for EPUB compatibility
     */
    private sanitizeXhtml(html: string): string {
        if (!html) return '';

        let safe = html;
        // Replace &nbsp; with numeric entity
        safe = safe.replace(/&nbsp;/g, '&#160;');
        // Remove empty paragraphs
        safe = safe.replace(/<p[^>]*>(\s|&nbsp;|&#160;|<br\s*\/?>)*<\/p>/gi, '');
        // Reduce excessive line breaks
        safe = safe.replace(/(<br\s*\/?>\s*){3,}/gi, '<br/><br/>');

        return safe.trim();
    }

    /**
     * Escape HTML special characters
     */
    private escapeHtml(text: string): string {
        return text
            .replace(/&/g, '&amp;')
            .replace(/</g, '&lt;')
            .replace(/>/g, '&gt;')
            .replace(/"/g, '&quot;')
            .replace(/'/g, '&#039;');
    }

    /**
     * Get the output path for an EPUB file based on build mode
     * - Merged + Original -> result/<BookName - Full>.epub
     * - Compressed (Any) -> result/<BookName>/compressed/<filename>
     * - Separate + Original -> result/<BookName>/original/<filename>
     */
    getOutputPath(filename: string, isMerged: boolean): string {
        const meta = this.metadata!;
        const bookNameSlug = formatFilename(meta.novelName);

        // Case 1: Merged & Original (Special Case)
        if (isMerged && !this.compressImages) {
            return join(this.outputDir, filename);
        }

        // Determine subfolder
        const subfolder = this.compressImages ? 'compressed' : 'original';

        // Path: result/<BookName>/<subfolder>/
        return join(this.outputDir, bookNameSlug, subfolder, filename);
    }

    /**
     * Process an image file with optional compression
     * Requirements: 6.4, 6.5, 6.6
     */
    async processImage(relPath: string): Promise<ProcessedImage | null> {
        if (!relPath) return null;

        // Check cache first
        if (this.imageCache.has(relPath)) {
            return this.imageCache.get(relPath)!;
        }

        const fullPath = join(this.baseFolder, relPath);

        // Check if file exists
        if (!(await fileExistsWithContent(fullPath))) {
            return null;
        }

        try {
            const imageBuffer = await readFile(fullPath);

            if (!this.compressImages) {
                // Preserve original format
                const ext = extname(relPath).toLowerCase();
                let mimeType = 'image/jpeg';
                if (ext === '.png') mimeType = 'image/png';
                else if (ext === '.gif') mimeType = 'image/gif';
                else if (ext === '.webp') mimeType = 'image/webp';

                const result: ProcessedImage = {
                    data: imageBuffer,
                    mimeType,
                    newPath: relPath,
                };
                this.imageCache.set(relPath, result);
                return result;
            }

            // Compress to JPEG quality 75
            try {
                const compressedBuffer = await sharp(imageBuffer)
                    .jpeg({ quality: 75 })
                    .toBuffer();

                // Change extension to .jpg
                const baseName = relPath.replace(/\.[^.]+$/, '');
                const newPath = `${baseName}.jpg`;

                const result: ProcessedImage = {
                    data: compressedBuffer,
                    mimeType: 'image/jpeg',
                    newPath,
                };
                this.imageCache.set(relPath, result);
                return result;
            } catch (sharpError) {
                // If sharp fails (unsupported format), return original as-is
                console.error(`\nFailed to compress image ${relPath}, using original:`, sharpError);
                const ext = extname(relPath).toLowerCase();
                let mimeType = 'image/jpeg';
                if (ext === '.png') mimeType = 'image/png';
                else if (ext === '.gif') mimeType = 'image/gif';
                else if (ext === '.webp') mimeType = 'image/webp';

                const result: ProcessedImage = {
                    data: imageBuffer,
                    mimeType,
                    newPath: relPath,
                };
                this.imageCache.set(relPath, result);
                return result;
            }
        } catch (error) {
            console.error(`\nFailed to process image ${relPath}:`, error);
            return null;
        }
    }


    /**
     * Create intro page with cover and metadata
     * Uses base64 data URIs for images
     */
    private async makeIntroPage(volName: string = ''): Promise<{ content: string; coverData: Buffer | null }> {
        const meta = this.metadata!;
        const summaryHtml = this.sanitizeXhtml(meta.summary);
        const title = this.escapeHtml(meta.novelName);
        const author = this.escapeHtml(meta.author);
        const tagsStr = meta.tags.join(', ');
        const tagsHtml = tagsStr ? `<p><b>Thể loại:</b> ${tagsStr}</p>` : '';

        let coverHtml = '<hr/>';
        let coverData: Buffer | null = null;

        if (meta.coverImageLocal) {
            const processed = await this.processImage(meta.coverImageLocal);
            if (processed) {
                coverData = processed.data;
                // Convert to base64 data URI
                const base64 = processed.data.toString('base64');
                const dataUri = `data:${processed.mimeType};base64,${base64}`;
                coverHtml = `<div style="text-align:center; margin: 2em 0; page-break-after: always; break-after: page;"><img src="${dataUri}" alt="Cover"/></div>`;
            }
        }

        const content = `
            <div style="text-align: center; margin-top: 5%;">
                <h1>${title}</h1>
                <h3 style="margin-bottom: 0.5em;">${volName}</h3>
                <p><b>Tác giả:</b> ${author}</p>
                ${tagsHtml}
                ${coverHtml}
                <div style="text-align: justify;">
                    ${summaryHtml}
                </div>
            </div>
        `;

        return { content, coverData };
    }

    /**
     * Process chapter content and extract/update image references
     * Converts images to base64 data URIs for epub-gen-memory compatibility
     */
    private async processChapterContent(content: string): Promise<{ html: string; images: ProcessedImage[] }> {
        const $ = cheerio.load(content, { xmlMode: false });
        const images: ProcessedImage[] = [];

        const imgElements = $('img').toArray();
        for (const img of imgElements) {
            const src = $(img).attr('src');
            if (src) {
                const processed = await this.processImage(src);
                if (processed) {
                    // Convert to base64 data URI for epub-gen-memory
                    const base64 = processed.data.toString('base64');
                    const dataUri = `data:${processed.mimeType};base64,${base64}`;
                    $(img).attr('src', dataUri);
                    images.push(processed);
                } else {
                    // Remove image if processing failed
                    $(img).remove();
                }
            }
        }

        return {
            html: this.sanitizeXhtml($.html()),
            images,
        };
    }

    /**
     * Build a merged EPUB containing all volumes
     * Requirements: 6.2, 6.7
     */
    async buildMerged(jsonFiles: string[]): Promise<string> {
        const meta = await this.loadMetadata();

        // Sort by volume order if available
        if (meta.volumes.length > 0) {
            const orderMap = new Map(meta.volumes.map((v) => [v.filename, v.order]));
            jsonFiles.sort((a, b) => (orderMap.get(a) ?? 9999) - (orderMap.get(b) ?? 9999));
        }

        // Create intro page
        const { content: introContent, coverData } = await this.makeIntroPage('Toàn tập');

        // Build chapters array for epub-gen-memory
        const chapters: EpubChapter[] = [];
        const allImages: Map<string, ProcessedImage> = new Map();

        // Add intro chapter
        chapters.push({
            title: 'Giới thiệu',
            content: introContent,
        });

        // Process each volume
        for (let volIndex = 0; volIndex < jsonFiles.length; volIndex++) {
            const jsonFile = jsonFiles[volIndex];
            console.log(`Merging: ${jsonFile}`);

            const volData = await this.loadVolumeData(jsonFile);

            // Create volume separator page
            let volHtmlContent = '';
            if (volData.coverImageLocal) {
                const processed = await this.processImage(volData.coverImageLocal);
                if (processed) {
                    allImages.set(processed.newPath, processed);
                    // Convert to base64 data URI
                    const base64 = processed.data.toString('base64');
                    const dataUri = `data:${processed.mimeType};base64,${base64}`;
                    volHtmlContent += `<img src="${dataUri}" alt="Vol Cover" style="max-height: 50vh;"/>`;
                }
            }
            volHtmlContent += `<h1>${this.escapeHtml(volData.volumeName)}</h1>`;

            const sepContent = `
                <div style="text-align: center; margin-top: 30vh;">
                    ${volHtmlContent}
                </div>
            `;

            chapters.push({
                title: volData.volumeName,
                content: sepContent,
            });

            // Process each chapter in the volume
            for (const chap of volData.chapters) {
                const { html, images } = await this.processChapterContent(chap.content);
                for (const img of images) {
                    allImages.set(img.newPath, img);
                }

                chapters.push({
                    title: chap.title,
                    content: `<h2>${this.escapeHtml(chap.title)}</h2>${html}`,
                });
            }
        }

        // Add cover image to allImages if exists
        if (coverData && meta.coverImageLocal) {
            const processed = await this.processImage(meta.coverImageLocal);
            if (processed) {
                allImages.set(processed.newPath, processed);
            }
        }

        // Build EPUB options
        const epubOptions: EpubGenOptions = {
            title: meta.novelName,
            author: meta.author,
            lang: 'vi',
            description: meta.summary,
            css: EPUB_CSS,
        };

        // Generate EPUB
        const epubBuffer = await epub(epubOptions, chapters);

        // Determine output path
        const baseFilename = formatFilename(`${meta.novelName} Full`);
        const filename = `${baseFilename}.epub`;
        const outputPath = this.getOutputPath(filename, true);

        // Ensure output directory exists and write file
        await ensureDir(join(outputPath, '..'));
        await writeFile(outputPath, epubBuffer);

        console.log(`Created Merged EPUB: ${outputPath}`);
        return outputPath;
    }


    /**
     * Build a single volume EPUB
     * Requirements: 6.3
     */
    async buildVolume(jsonFile: string): Promise<string> {
        const meta = await this.loadMetadata();
        const volData = await this.loadVolumeData(jsonFile);

        // Create intro page with volume name
        const { content: introContent, coverData } = await this.makeIntroPage(volData.volumeName);

        // Build chapters array
        const chapters: EpubChapter[] = [];
        const allImages: Map<string, ProcessedImage> = new Map();

        // Add intro chapter
        chapters.push({
            title: 'Giới thiệu',
            content: introContent,
        });

        // Process each chapter
        for (const chap of volData.chapters) {
            const { html, images } = await this.processChapterContent(chap.content);
            for (const img of images) {
                allImages.set(img.newPath, img);
            }

            chapters.push({
                title: chap.title,
                content: `<h2>${this.escapeHtml(chap.title)}</h2>${html}`,
            });
        }

        // Add cover image if exists
        if (coverData && meta.coverImageLocal) {
            const processed = await this.processImage(meta.coverImageLocal);
            if (processed) {
                allImages.set(processed.newPath, processed);
            }
        }

        // Build EPUB options
        const epubOptions: EpubGenOptions = {
            title: `${volData.volumeName} - ${meta.novelName}`,
            author: meta.author,
            lang: 'vi',
            css: EPUB_CSS,
        };

        // Generate EPUB
        const epubBuffer = await epub(epubOptions, chapters);

        // Determine output path
        const baseFilename = formatFilename(`${volData.volumeName} - ${meta.novelName}`);
        const filename = `${baseFilename}.epub`;
        const outputPath = this.getOutputPath(filename, false);

        // Ensure output directory exists and write file
        await ensureDir(join(outputPath, '..'));
        await writeFile(outputPath, epubBuffer);

        console.log(`Created: ${outputPath}`);
        return outputPath;
    }

    /**
     * Clear the image cache
     */
    clearCache(): void {
        this.imageCache.clear();
    }
}
